gtn/functions/compose.cpp

```diff
@@ -450,3 +456,931 @@ Graph compose(
 
 } // namespace detail
 } // namespace gtn
+
+namespace gtn {
+namespace detail {
+namespace dataparallel {
+// Change AOS to SOA
+struct GraphDataParallel {
+  // True if a node is accept or start, false otherwise
+  std::vector<bool> accept;
+  std::vector<bool> start;
+
+  // One value per node - i-th value corresponds to i-th node
+  // Last element is the total number of arcs, so that
+  // each element and its neighbor forms a range
+  std::vector<int> inArcOffset;
+  std::vector<int> outArcOffset;
+
+  // One value per arc
+  std::vector<int> inArcs;
+  std::vector<int> outArcs;
+
+  // One value per arc
+  // i-th value corresponds to i-th arc
+  std::vector<int> ilabels;
+  std::vector<int> olabels;
+  std::vector<int> srcNodes;
+  std::vector<int> dstNodes;
+  std::vector<float> weights;
+};
+
+namespace {
+// Exclusive/Inclusive prefix sum. The returned vector
+// has one more element
+int prefixSumScan(std::vector<int>& input, bool appendSum) {
+  int sum = 0;
+  for (size_t i = 0; i < input.size(); ++i) {
+    auto count = input[i];
+    input[i] = sum;
+    sum += count;
+  }
+  if (appendSum) {
+    input.push_back(sum);
+  }
+
+  return sum;
+}
+
+// Remove all nodes that have 0 in and out arcs - these nodes are unreachable
+// This is essentially a stream compact
+std::tuple<std::vector<int>, std::vector<int>> removeUnreachableNodes(
+    const std::vector<int>& numInArcs,
+    const std::vector<int>& numOutArcs) {
+  std::vector<int> outputInArcs;
+  std::vector<int> outputOutArcs;
+
+  assert(numInArcs.size() == numOutArcs.size());
+
+  for (size_t i = 0; i < numInArcs.size(); ++i) {
+    if ((numInArcs[i] != 0) || (numOutArcs[i] != 0)) {
+      outputInArcs.push_back(numInArcs[i]);
+      outputOutArcs.push_back(numOutArcs[i]);
+    }
+  }
+
+  return std::make_pair(outputInArcs, outputOutArcs);
+}
+
+// TODO: Duplicate - should be removed
+inline int TwoDToOneDIndex(int n1, int n2, int n1Extent) {
```

Comment:
Do you need to check that `n1 < n1Extent`?

Reply:
Thanks!! will add `assert`